### PR TITLE
Checkout & test future merge when adding "safe to test"

### PR DESCRIPTION
**Pull request by cwli24**
### Overview
Rectify what's being tested on PRs when adding the test label.

Note that per-discussion, this does not include consideration of merge conflicts. Adding the test label to a PR with a merge conflict (unable to merge) is both wasteful and unknown behavior. At this time, ensuring the validity of passing test checks is manually handled.

### Related Github and/or JIRA Issue(s)
NR-72380

### Testing
Workflow dispatches should still be checking out specified branch heads.
PR test label should now be checking out the "future" merge of its latest and its base's latest, on mergeable PR state.

### CODE REVIEW

File: .github/reusable_wfs/setup.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        id: evt
+        run: |
+          if [ ${{ github.event_name }} == 'pull_request_target' ]; then
+            echo "::set-output name=ref::refs/pull/${{ github.event.number }}/merge"
+          elif [ ${{ github.event_name }} == 'workflow_dispatch' ]; then
+            echo "::set-output name=ref::${{ github.event.ref }}"
+          fi
+        shell: bash
+
+      - uses: actions/checkout@v2
```
Comment (patrickhousley) [inline .github/reusable_wfs/setup.yml:22]: Can this be updated to use the latest version of the action?

```suggestion
      - uses: actions/checkout@v3
```

**Comment by patrickhousley** [comment]
Just looking through the GHA docs, it looks like re-usable actions is stored in `.github/actions/setup.yml`. Maybe rename `reusable_wfs` to `actions`

**Comment by patrickhousley** [comment]
We could also take this opportunity to better define the workflows so things like the build and node modules are shared across the jobs instead of each job duplication effort.

**Comment by cwli24** [comment]
> We could also take this opportunity to better define the workflows so things like the build and node modules are shared across the jobs instead of each job duplication effort.

I thought about this; what I got was it'd be downloading artifacts to share things across jobs. Is that worth it given our checkout & build isn't that heavy?

**Comment by patrickhousley** [comment]
> I thought about this; what I got was it'd be downloading artifacts to share things across jobs. Is that worth it given our checkout & build isn't that heavy?

The npm install and build take roughly 2 minutes in each job. We could probably save at least a little bit of time.

**Comment by cwli24** [comment]
> specified

Is it necessary to run, or can you have different folders (name) for these workflows?

**Comment by patrickhousley** [comment]
> > Just looking through the GHA docs, it looks like re-usable actions is stored in `.github/actions/setup.yml`. Maybe rename `reusable_wfs` to `actions`
> 
> Is it necessary to run, or can you have different folders (name) for these workflows?

The only requirement is that workflows that are executable by GHA, either automatically or manually, exist in `.githun/workflows`. Shared actions and workflows can exist anywhere else.

**Comment by cwli24** [comment]
> > > Just looking through the GHA docs, it looks like re-usable actions is stored in `.github/actions/setup.yml`. Maybe rename `reusable_wfs` to `actions`
> > 
> > 
> > Is it necessary to run, or can you have different folders (name) for these workflows?
> 
> The only requirement is that workflows that are executable by GHA, either automatically or manually, exist in `.githun/workflows`. Shared actions and workflows can exist anywhere else.

Then I still think having a reusable_wf is more descriptive and intuitive than "action", agree/disagree?

**Comment by metal-messiah** [comment]
> 
> 
> 
> 
> 
> 
> > > > Just looking through the GHA docs, it looks like re-usable actions is stored in `.github/actions/setup.yml`. Maybe rename `reusable_wfs` to `actions`
> 
> > > 
> 
> > > 
> 
> > > Is it necessary to run, or can you have different folders (name) for these workflows?
> 
> > 
> 
> > The only requirement is that workflows that are executable by GHA, either automatically or manually, exist in `.githun/workflows`. Shared actions and workflows can exist anywhere else.
> 
> 
> 
> Then I still think having a reusable_wf is more descriptive and intuitive than "action", agree/disagree?

I'm fine with either. The part to me that's vague is "setup", which is generic, but inside the action the name says it's specifically for tests (and runs build:all as opposed to build:prod for example). So this isn't a generic action and the name should probably reflect it's meant for tests just for brevity.

File: .github/workflows/tests-polyfill.yml [hunk 2]
```diff
@@ -20,13 +20,7 @@ jobs:
...
 
     steps:
-      - uses: actions/checkout@v2
-
-      - name: install
-        run: npm ci --cache ./.npm
-
-      - name: build
-        run: npm run build:all
+      - uses: ./.github/reusable_wfs/test-setup.yml
```
[the remark this answers is not in the document]
Comment (cwli24) [reply .github/workflows/tests-polyfill.yml:23]: I'm not understand

**Review by patrickhousley — APPROVED** [review]

**Review by metal-messiah — APPROVED** [review]

**Comment by cwli24** [comment]
Update: this PR has been reverted due to failures. Will re-open and retry with composite actions rather than reusable workflow.

**Pull request merged**